open-craft/opencraft
# [BB-2008] Migrate to Gandi v5 LiveDNS API

This PR adds the code changes required to use the Gandi v5 LiveDNS API to manage the DNS records of the Ocim domains.

**Update**:

This also contains the fixes to:
* fix the failing integration cleanup job.
* fix the `xqueue` service error breaking provisioning.

**Testing instructions**:
* Ensure that the CI checks pass and there are no errors in the integration tests related to managing the DNS records.
* Check out the source branch of this PR and run Ocim.
* Create a new instance and spawn a new appserver for it.
* Check and verify that the DNS records for the instance are created during provisioning and the added records are present in the zone file for the domain in the Gandi admin interface. After a short while, ensure that the domains in the added records resolve properly.
* Once the instance has completed provisioning, verify that the DNS record for the active appserver VM is added without any errors.
* Archive the instance and verify that the DNS records are removed successfully.

**Testing instructions for other fixes**:
* Verify that the integration cleanup job doesn't fail due to an error about not being able to find the `coverage` command.
* Verify that the instance provisioning completes successfully and the right version of `xqueue` is installed.

**Author notes and concerns**:
* The API key for the account with access to the integration test base domain, doesn't have the permission to list the domains in the account. So the value  of the `DEFAULT_INSTANCE_BASE_DOMAIN` environment variable is added as a fallback. We have to ensure that the API key used in production has all the required permissions.
* The code for the existing XML-RPC API is left intact and can be restored easily with very minimal changes. This can be removed after the migration in production is complete and verified to work without any issues.
* The `cleanup_utils` code changes should work, butt the actual removal of the records couldn't be tested locally due to there being no data left to cleanup.
* This PR has to be merged very close to the deployment on production, after all the preparation for the domain migration is done and a maintenance window for Ocim is announced.
